rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+//! VAPID auth support
+//!
+//! This library only supports the latest VAPID-draft-02+ specification.
+//!
+//! Example Use:
+//! ```
+//!
+//! // Create a key from an existing EC Private Key PEM file.
+//! // You can generate this with
+//! // Key::generate().to_pem("pem/file/path.pem")?;
+//! let my_key = Key::from_pem("pem/file/path.pem")?;
+//!
+//! // Construct the Claims hashmap
+//! let mut claims:HashMap<String, serde_json::Value> = HashMap::new();
+//! claims.insert(
+//!     String::from("sub"), serde_json::Value::from("mailto:bob@example.com")
+//! );
+//! // while `exp` can be filled in for you, `aud` should point to the net location of the
+//! // Push server you wish to talk to. (e.g. `https://push.services.mozilla.org`)
+//! // `aud` is optional for Mozilla, but may be required for GCM/FCM or other systems.
+//! claims.insert(
+//!     String::from("aud"), serde_json::Value::from("https://host.ext")
+//! );
+//!
+//! // The result will contain the `Authorization:` header. How you inject this into your
+//! // request is left as an exercise.
+//! let authorization_header = sign(key, &mut claims)?;
+//!
+//! ```
+
+extern crate asn1;
+extern crate base64;
+extern crate openssl;
+extern crate serde_json;
+extern crate time;
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::io::Write;
+use std::path::Path;
+
+use openssl::bn::BigNum;
+use openssl::bn::BigNumContext;
+use openssl::hash::MessageDigest;
+use openssl::ec;
+use openssl::nid;
+use openssl::sign::{Signer, Verifier};
+use openssl::pkey;
+
+pub struct Key(ec::EcKey);
+
+/// a Key is a helper for creating or using a VAPID EC key.
+///
+/// Vapid Keys are always Prime256v1 EC keys.
+///
+#[allow(dead_code)]
```

Comment:
Was this needed to get things compiling? Or just used during development?

Unfortunately we don't have a great story today for "don't warn locally and error/warn on CI", although it can be manually constructed through a number of means if desired.

Reply:
Ah, it was indeed to get things rolling (and to get pycharm to shut up a little). Will remove.